packages/components/src/query-controls/types.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * External dependencies
+ */
+import type { Key } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import type { FormTokenFieldProps } from '../form-token-field/types';
+import type { TreeSelectProps } from '../tree-select/types';
+
+export type Entity = {
+	id: number;
+	name: string;
+	parent?: number | null;
+	value?: string;
+};
+
+export type Category = {
+	id: number;
+	value: string;
+	name?: string;
+};
+
+export type TermsWithChildren = Array< Entity & { children: Entity[] } >;
+
+export type CategorySelectProps = {
+	categoriesList: QueryControlsProps[ 'categoriesList' ];
+	key: Key;
+	label: TreeSelectProps[ 'label' ];
+	noOptionLabel: TreeSelectProps[ 'noOptionLabel' ];
+	onChange: TreeSelectProps[ 'onChange' ];
+	selectedCategoryId: QueryControlsProps[ 'selectedCategoryId' ];
+};
+
+export type AuthorSelectProps = {
+	authorList: QueryControlsProps[ 'authorList' ];
+	label: TreeSelectProps[ 'label' ];
+	noOptionLabel: TreeSelectProps[ 'noOptionLabel' ];
+	onChange: QueryControlsProps[ 'onAuthorChange' ];
+	selectedAuthorId: QueryControlsProps[ 'selectedAuthorId' ];
+};
+
+export type QueryControlsProps = {
+	/**
+	 * An array of authors that is passed into
+	 * an `AuthorSelect` sub-component.
+	 */
+	authorList?: Entity[];
+	/**
+	 * An array of categories, renders a `CategorySelect`
+	 * sub-component when passed in conjunction with `onCategoryChange`.
+	 */
+	categoriesList?: Entity[];
+	/**
+	 * An array of category names, renders a `FormTokenField` component
+	 * when passed in conjunction with `onCategoryChange`.
+	 */
+	categorySuggestions?: { [ categoryName: Entity[ 'name' ] ]: Entity };
+	/**
+	 * The maximum items.
+	 *
+	 * @default 100
+	 */
+	maxItems?: number;
+	/**
+	 * The minimum of items.
+	 *
+	 * @default 1
+	 */
+	minItems?: number;
+	/**
+	 * The selected number of items to retrieve via the query.
+	 */
+	numberOfItems?: number;
+	/**
+	 * A function that receives the new author value.
+	 * If this is not specified, the author controls are not included.
+	 */
+	onAuthorChange?: TreeSelectProps[ 'onChange' ];
+	/**
+	 * A function that receives the new category value.
+	 * If this is not specified, the category controls are not included.
+	 */
+	onCategoryChange?: TreeSelectProps[ 'onChange' ] &
```

Comment:
Yeah, I wanted to do `|` instead, but there were these errors:

```
packages/components/src/query-controls/index.tsx(114,7): error TS2322: Type '((value: string | string[], extra?: { event?: ChangeEvent<HTMLSelectElement> | undefined; } | undefined) => void) | ((tokens: (string | TokenItem)[]) => void)' is not assignable to type '(value: string | string[], extra?: { event?: ChangeEvent<HTMLSelectElement> | undefined; } | undefined) => void'.
Type '(tokens: (string | TokenItem)[]) => void' is not assignable to type '(value: string | string[], extra?: { event?: ChangeEvent<HTMLSelectElement> | undefined; } | undefined) => void'.
  Types of parameters 'tokens' and 'value' are incompatible.
    Type 'string | string[]' is not assignable to type '(string | TokenItem)[]'.
      Type 'string' is not assignable to type '(string | TokenItem)[]'.
packages/components/src/query-controls/index.tsx(129,7): error TS2322: Type '((value: string | string[], extra?: { event?: ChangeEvent<HTMLSelectElement> | undefined; } | undefined) => void) | ((tokens: (string | TokenItem)[]) => void)' is not assignable to type '((tokens: (string | TokenItem)[]) => void) | undefined'.
Type '(value: string | string[], extra?: { event?: ChangeEvent<HTMLSelectElement> | undefined; } | undefined) => void' is not assignable to type '(tokens: (string | TokenItem)[]) => void'.
  Types of parameters 'value' and 'tokens' are incompatible.
    Type '(string | TokenItem)[]' is not assignable to type 'string | string[]'.
      Type '(string | TokenItem)[]' is not assignable to type 'string[]'.
        Type 'string | TokenItem' is not assignable to type 'string'.
          Type 'TokenItem' is not assignable to type 'string'.
```

<img width="944" alt="Screen Shot 2023-01-12 at 4 41 38 PM" src="https://user-images.githubusercontent.com/4063887/212197051-b758cee1-cb81-400b-845b-0e8d8a3fd21a.png">

Reply:
Still, `&` seems hacky

Reply:
Yeah, that doesn't look like the correct way to represent those types.

I decided to take a deeper look at this, and ended up in a bit of a rabbit hole 🐰 

By reading the docs, it looks like `QueryControls` can be used with single or multiple category selection depending on which props are passed to the component. Based off that, the `onCategoryChange` prop changes its type (either `TreeSelectProps[ 'onChange' ]` for single selection, or `FormTokenFieldProps[ 'onChange' ]` for multiple selection).

The correct way to represent this duality in TypeScript is to split the "single selection props" from the "multiple selection props". Something like this (pseudo code):

```ts
type CommonProps = { ... };
type WithSingleSelection = CommonProps & {
  // props specific to single category selection
  onChange?: TreeSelectProps[ 'onChange' ];
};
type WithMultipleSelection = CommonProps & {
  // props specific to multiple category selection
  onChange?: FormTokenFieldProps[ 'onChange' ];
}; 
type QueryControlsProps = WithSingleSelection | WithMultipleSelection;
```

This can be paired, in the `index.tsx` file, with type guards, which basically allow us to tell TypeScript to consider the incoming `props` as single or multiple selection. (pseudo code): 

```ts
function isMultipleCategorySelection(
	props: QueryControlsProps
): props is WithMultipleSelection {
	return 'categorySuggestions' in props;
}


// later in code
		{
					isMultipleCategorySelection( props ) &&
					props.categorySuggestions &&
					props.onCategoryChange && (
						<FormTokenField
```

I then realized that using `TreeSelectProps[ 'onChange' ];` as the type for the `onChange` prop in the `CategorySelect` and `AuthorSelect` component is not necessary. We can instead type it as a simpler `( newValue: string ) => void` function, because:
- the `newValue` will never be `string[]` (because that only happens when passing the `multiple` prop to `TreeSelect`
- the second argument has never been assumed to exist for consumers of the component

I also decided to swap the type "dependency" from `CategorySelectProps` and `AuthorSelectProps` to `QueryControlsProps` — I think it's better if the prop types are defined in `CategorySelectProps` and `AuthorSelectProps` as those components defines the implementation details, and if the type definitions get reused in `QueryControlsProps`

Finally, I started testing those changes in the Storybook example, and I decided to make some tweaks:
- I realized that all `on.*` callbacks were not being shown in the `actions` tab because they were not being passed explicitly to the component. I went ahead and fixed that.
- I tweaked the `handleCategoryChange` function, since I noticed that the type cast to `Category[]` was actually hiding an edge case that would have caused runtime errors.

<details>

<summary>All together, these are the changes that I applied:</summary>

```diff
diff --git a/packages/components/src/query-controls/author-select.tsx b/packages/components/src/query-controls/author-select.tsx
index 37163aba4d..ebda05ffb9 100644
--- a/packages/components/src/query-controls/author-select.tsx
+++ b/packages/components/src/query-controls/author-select.tsx
@@ -3,6 +3,7 @@
  */
 import { buildTermsTree } from './terms';
 import TreeSelect from '../tree-select';
+import type { TreeSelectProps } from '../tree-select/types';
 import type { AuthorSelectProps } from './types';
 
 export default function AuthorSelect( {
@@ -10,13 +11,21 @@ export default function AuthorSelect( {
 	noOptionLabel,
 	authorList,
 	selectedAuthorId,
-	onChange,
+	onChange: onChangeProp,
 }: AuthorSelectProps ) {
 	if ( ! authorList ) return null;
 	const termsTree = buildTermsTree( authorList );
 	return (
 		<TreeSelect
-			{ ...{ label, noOptionLabel, onChange } }
+			{ ...{
+				label,
+				noOptionLabel,
+				// Since the `multiple` attribute is not passed to `TreeSelect`, it is
+				// safe to assume that the argument of `onChange` cannot be `string[]`.
+				// The correct solution would be to type `SelectControl` better, so that
+				// the type of `value` and `onChange` vary depending on `multiple`.
+				onChange: onChangeProp as TreeSelectProps[ 'onChange' ],
+			} }
 			tree={ termsTree }
 			selectedId={
 				selectedAuthorId !== undefined
diff --git a/packages/components/src/query-controls/category-select.tsx b/packages/components/src/query-controls/category-select.tsx
index c082e3d977..938dfe792d 100644
--- a/packages/components/src/query-controls/category-select.tsx
+++ b/packages/components/src/query-controls/category-select.tsx
@@ -3,6 +3,7 @@
  */
 import { buildTermsTree } from './terms';
 import TreeSelect from '../tree-select';
+import type { TreeSelectProps } from '../tree-select/types';
 
 /**
  * WordPress dependencies
@@ -15,7 +16,7 @@ export default function CategorySelect( {
 	noOptionLabel,
 	categoriesList,
 	selectedCategoryId,
-	onChange,
+	onChange: onChangeProp,
 	...props
 }: CategorySelectProps ) {
 	const termsTree = useMemo( () => {
@@ -24,7 +25,15 @@ export default function CategorySelect( {
 
 	return (
 		<TreeSelect
-			{ ...{ label, noOptionLabel, onChange } }
+			{ ...{
+				label,
+				noOptionLabel,
+				// Since the `multiple` attribute is not passed to `TreeSelect`, it is
+				// safe to assume that the argument of `onChange` cannot be `string[]`.
+				// The correct solution would be to type `SelectControl` better, so that
+				// the type of `value` and `onChange` vary depending on `multiple`.
+				onChange: onChangeProp as TreeSelectProps[ 'onChange' ],
+			} }
 			tree={ termsTree }
 			selectedId={
 				selectedCategoryId !== undefined
diff --git a/packages/components/src/query-controls/index.tsx b/packages/components/src/query-controls/index.tsx
index 4cacc06d76..f940540963 100644
--- a/packages/components/src/query-controls/index.tsx
+++ b/packages/components/src/query-controls/index.tsx
@@ -11,12 +11,28 @@ import CategorySelect from './category-select';
 import FormTokenField from '../form-token-field';
 import RangeControl from '../range-control';
 import SelectControl from '../select-control';
-import type { QueryControlsProps } from './types';
+import type {
+	QueryControlsProps,
+	QueryControlsWithMultipleCategorySelectionProps,
+	QueryControlsWithSingleCategorySelectionProps,
+} from './types';
 
 const DEFAULT_MIN_ITEMS = 1;
 const DEFAULT_MAX_ITEMS = 100;
 const MAX_CATEGORIES_SUGGESTIONS = 20;
 
+function isSingleCategorySelection(
+	props: QueryControlsProps
+): props is QueryControlsWithSingleCategorySelectionProps {
+	return 'categoriesList' in props;
+}
+
+function isMultipleCategorySelection(
+	props: QueryControlsProps
+): props is QueryControlsWithMultipleCategorySelectionProps {
+	return 'categorySuggestions' in props;
+}
+
 /**
  * Controls to query for posts.
  *
@@ -45,20 +61,18 @@ const MAX_CATEGORIES_SUGGESTIONS = 20;
 export function QueryControls( {
 	authorList,
 	selectedAuthorId,
-	categoriesList,
-	selectedCategoryId,
-	categorySuggestions,
-	selectedCategories,
 	numberOfItems,
 	order,
 	orderBy,
 	maxItems = DEFAULT_MAX_ITEMS,
 	minItems = DEFAULT_MIN_ITEMS,
-	onCategoryChange,
 	onAuthorChange,
 	onNumberOfItemsChange,
 	onOrderChange,
 	onOrderByChange,
+	// Props for single OR multiple category selection are not destructured here,
+	// but instead are destructured inline where necessary.
+	...props
 }: QueryControlsProps ) {
 	return (
 		<>
@@ -112,32 +126,38 @@ export function QueryControls( {
 						} }
 					/>
 				),
-				categoriesList && onCategoryChange && (
-					<CategorySelect
-						key="query-controls-category-select"
-						categoriesList={ categoriesList }
-						label={ __( 'Category' ) }
-						noOptionLabel={ __( 'All' ) }
-						selectedCategoryId={ selectedCategoryId }
-						onChange={ onCategoryChange }
-					/>
-				),
-				categorySuggestions && onCategoryChange && (
-					<FormTokenField
-						key="query-controls-categories-select"
-						label={ __( 'Categories' ) }
-						value={
-							selectedCategories &&
-							selectedCategories.map( ( item ) => ( {
-								id: item.id,
-								value: item.name || item.value,
-							} ) )
-						}
-						suggestions={ Object.keys( categorySuggestions ) }
-						onChange={ onCategoryChange }
-						maxSuggestions={ MAX_CATEGORIES_SUGGESTIONS }
-					/>
-				),
+				isSingleCategorySelection( props ) &&
+					props.categoriesList &&
+					props.onCategoryChange && (
+						<CategorySelect
+							key="query-controls-category-select"
+							categoriesList={ props.categoriesList }
+							label={ __( 'Category' ) }
+							noOptionLabel={ __( 'All' ) }
+							selectedCategoryId={ props.selectedCategoryId }
+							onChange={ props.onCategoryChange }
+						/>
+					),
+				isMultipleCategorySelection( props ) &&
+					props.categorySuggestions &&
+					props.onCategoryChange && (
+						<FormTokenField
+							key="query-controls-categories-select"
+							label={ __( 'Categories' ) }
+							value={
+								props.selectedCategories &&
+								props.selectedCategories.map( ( item ) => ( {
+									id: item.id,
+									value: item.name || item.value,
+								} ) )
+							}
+							suggestions={ Object.keys(
+								props.categorySuggestions
+							) }
+							onChange={ props.onCategoryChange }
+							maxSuggestions={ MAX_CATEGORIES_SUGGESTIONS }
+						/>
+					),
 				onAuthorChange && (
 					<AuthorSelect
 						key="query-controls-author-select"
diff --git a/packages/components/src/query-controls/stories/index.tsx b/packages/components/src/query-controls/stories/index.tsx
index e62b1442c1..00478e5b0d 100644
--- a/packages/components/src/query-controls/stories/index.tsx
+++ b/packages/components/src/query-controls/stories/index.tsx
@@ -12,7 +12,12 @@ import { useState } from '@wordpress/element';
  * Internal dependencies
  */
 import QueryControls from '..';
-import type { Category, QueryControlsProps } from '../types';
+import type {
+	Category,
+	QueryControlsProps,
+	QueryControlsWithSingleCategorySelectionProps,
+	QueryControlsWithMultipleCategorySelectionProps,
+} from '../types';
 
 const meta: ComponentMeta< typeof QueryControls > = {
 	title: 'Components/QueryControls',
@@ -33,7 +38,16 @@ const meta: ComponentMeta< typeof QueryControls > = {
 };
 export default meta;
 
-const DefaultTemplate: ComponentStory< typeof QueryControls > = ( props ) => {
+export const Default: ComponentStory< typeof QueryControls > = ( args ) => {
+	const {
+		onAuthorChange,
+		onCategoryChange,
+		onNumberOfItemsChange,
+		onOrderByChange,
+		onOrderChange,
+		...props
+	} = args as QueryControlsWithMultipleCategorySelectionProps;
+
 	const [ ownNumberOfItems, setOwnNumberOfItems ] = useState(
 		props.numberOfItems
 	);
@@ -46,37 +60,51 @@ const DefaultTemplate: ComponentStory< typeof QueryControls > = ( props ) => {
 		props.selectedCategories
 	);
 
-	const onCategoryChange: QueryControlsProps[ 'onCategoryChange' ] = (
-		tokens
-	) => {
-		if ( 'string' === typeof tokens ) {
-			return;
-		}
+	const handleCategoryChange: QueryControlsWithMultipleCategorySelectionProps[ 'onCategoryChange' ] =
+		( tokens ) => {
+			onCategoryChange?.( tokens );
+
+			const hasNoSuggestion = tokens.some(
+				( token ) =>
+					typeof token === 'string' &&
+					! props.categorySuggestions?.[ token ]
+			);
+			if ( hasNoSuggestion ) {
+				return;
+			}
 
-		const allCategories = tokens?.map( ( token ) => {
-			return typeof token === 'string'
-				? props.categorySuggestions?.[ token ]
-				: token;
-		} ) as Category[];
+			const allCategories = tokens
+				.map( ( token ) => {
+					return typeof token === 'string'
+						? props.categorySuggestions?.[ token ]
+						: token;
+				} )
+				.filter( Boolean ) as Category[];
 
-		setOwnSelectedCategories( allCategories );
-	};
+			setOwnSelectedCategories( allCategories );
+		};
 
 	return (
 		<QueryControls
 			{ ...props }
 			numberOfItems={ ownNumberOfItems }
-			onCategoryChange={ onCategoryChange }
+			onCategoryChange={ handleCategoryChange }
 			onOrderByChange={ ( newOrderBy ) => {
+				onOrderByChange?.( newOrderBy );
 				setOwnOrderBy( newOrderBy );
 			} }
 			onOrderChange={ ( newOrder ) => {
+				onOrderChange?.( newOrder );
 				setOwnOrder( newOrder );
 			} }
 			order={ ownOrder }
 			orderBy={ ownOrderBy }
-			onNumberOfItemsChange={ setOwnNumberOfItems }
+			onNumberOfItemsChange={ ( newNumber ) => {
+				onNumberOfItemsChange?.( newNumber );
+				setOwnNumberOfItems( newNumber );
+			} }
 			onAuthorChange={ ( newAuthor ) => {
+				onAuthorChange?.( newAuthor );
 				setOwnSelectedAuthorId( Number( newAuthor ) );
 			} }
 			selectedAuthorId={ ownSelectedAuthorId }
@@ -84,9 +112,6 @@ const DefaultTemplate: ComponentStory< typeof QueryControls > = ( props ) => {
 		/>
 	);
 };
-
-export const Default: ComponentStory< typeof QueryControls > =
-	DefaultTemplate.bind( {} );
 Default.args = {
 	authorList: [
 		{
@@ -125,25 +150,38 @@ Default.args = {
 };
 
 const SingleCategoryTemplate: ComponentStory< typeof QueryControls > = (
-	props
+	args
 ) => {
+	const {
+		onAuthorChange,
+		onCategoryChange,
+		onNumberOfItemsChange,
+		onOrderByChange,
+		onOrderChange,
+		...props
+	} = args as QueryControlsWithSingleCategorySelectionProps;
 	const [ ownOrder, setOwnOrder ] = useState( props.order );
 	const [ ownOrderBy, setOwnOrderBy ] = useState( props.orderBy );
 	const [ ownSelectedCategoryId, setSelectedCategoryId ] = useState(
 		props.selectedCategoryId
 	);
 
+	const handleCategoryChange: QueryControlsWithSingleCategorySelectionProps[ 'onCategoryChange' ] =
+		( newCategory ) => {
+			onCategoryChange?.( newCategory );
+			setSelectedCategoryId( Number( newCategory ) );
+		};
+
 	return (
 		<QueryControls
 			{ ...props }
-			onCategoryChange={ ( newCategory ) => {
-				onCategoryChange?.( newCategory );
-				setSelectedCategoryId( Number( newCategory ) );
-			} }
+			onCategoryChange={ handleCategoryChange }
 			onOrderByChange={ ( newOrderBy ) => {
+				onOrderByChange?.( newOrderBy );
 				setOwnOrderBy( newOrderBy as QueryControlsProps[ 'orderBy' ] );
 			} }
 			onOrderChange={ ( newOrder ) => {
+				onOrderChange?.( newOrder );
 				setOwnOrder( newOrder as QueryControlsProps[ 'order' ] );
 			} }
 			order={ ownOrder }
diff --git a/packages/components/src/query-controls/types.ts b/packages/components/src/query-controls/types.ts
index bf169057b8..34e256ef87 100644
--- a/packages/components/src/query-controls/types.ts
+++ b/packages/components/src/query-controls/types.ts
@@ -1,8 +1,3 @@
-/**
- * External dependencies
- */
-import type { Key } from 'react';
-
 /**
  * Internal dependencies
  */
@@ -33,39 +28,30 @@ export type Category = {
 type Order = 'asc' | 'desc';
 type OrderBy = 'date' | 'title';
 
-export type CategorySelectProps = {
-	categoriesList: NonNullable< QueryControlsProps[ 'categoriesList' ] >;
-	key: Key;
-	label: TreeSelectProps[ 'label' ];
-	noOptionLabel: TreeSelectProps[ 'noOptionLabel' ];
-	onChange: NonNullable< TreeSelectProps[ 'onChange' ] >;
-	selectedCategoryId: QueryControlsProps[ 'selectedCategoryId' ];
+export type CategorySelectProps = Pick<
+	TreeSelectProps,
+	'label' | 'noOptionLabel'
+> & {
+	categoriesList: Entity[];
+	onChange: ( newCategory: string ) => void;
+	selectedCategoryId?: Category[ 'id' ];
 };
 
-export type AuthorSelectProps = {
-	authorList: QueryControlsProps[ 'authorList' ];
-	label: TreeSelectProps[ 'label' ];
-	noOptionLabel: TreeSelectProps[ 'noOptionLabel' ];
-	onChange: NonNullable< QueryControlsProps[ 'onAuthorChange' ] >;
-	selectedAuthorId: QueryControlsProps[ 'selectedAuthorId' ];
+export type AuthorSelectProps = Pick<
+	TreeSelectProps,
+	'label' | 'noOptionLabel'
+> & {
+	authorList?: Entity[];
+	onChange: ( newAuthor: string ) => void;
+	selectedAuthorId?: Entity[ 'id' ];
 };
 
-export type QueryControlsProps = {
+type BaseQueryControlsProps = {
 	/**
 	 * An array of authors that is passed into
 	 * an `AuthorSelect` sub-component.
 	 */
-	authorList?: Entity[];
-	/**
-	 * An array of categories, renders a `CategorySelect`
-	 * sub-component when passed in conjunction with `onCategoryChange`.
-	 */
-	categoriesList?: Entity[];
-	/**
-	 * An array of category names, renders a `FormTokenField` component
-	 * when passed in conjunction with `onCategoryChange`.
-	 */
-	categorySuggestions?: { [ categoryName: Entity[ 'name' ] ]: Entity };
+	authorList?: AuthorSelectProps[ 'authorList' ];
 	/**
 	 * The maximum items.
 	 *
@@ -86,13 +72,7 @@ export type QueryControlsProps = {
 	 * A function that receives the new author value.
 	 * If this is not specified, the author controls are not included.
 	 */
-	onAuthorChange?: TreeSelectProps[ 'onChange' ];
-	/**
-	 * A function that receives the new category value.
-	 * If this is not specified, the category controls are not included.
-	 */
-	onCategoryChange?: TreeSelectProps[ 'onChange' ] &
-		FormTokenFieldProps[ 'onChange' ];
+	onAuthorChange?: AuthorSelectProps[ 'onChange' ];
 	/**
 	 * A function that receives the new number of items value.
 	 * If this is not specified, then the number of items
@@ -122,13 +102,45 @@ export type QueryControlsProps = {
 	/**
 	 * The selected author ID.
 	 */
-	selectedAuthorId?: Entity[ 'id' ];
-	/**
-	 * The selected categories for the `categorySuggestions`.
-	 */
-	selectedCategories?: Category[];
-	/**
-	 * The selected category for the `categoriesList`.
-	 */
-	selectedCategoryId?: Category[ 'id' ];
+	selectedAuthorId?: AuthorSelectProps[ 'selectedAuthorId' ];
 };
+
+export type QueryControlsWithSingleCategorySelectionProps =
+	BaseQueryControlsProps & {
+		/**
+		 * An array of categories, renders a `CategorySelect`
+		 * sub-component when passed in conjunction with `onCategoryChange`.
+		 */
+		categoriesList?: CategorySelectProps[ 'categoriesList' ];
+		/**
+		 * The selected category for the `categoriesList`.
+		 */
+		selectedCategoryId?: CategorySelectProps[ 'selectedCategoryId' ];
+		/**
+		 * A function that receives the new category value.
+		 * If this is not specified, the category controls are not included.
+		 */
+		onCategoryChange?: CategorySelectProps[ 'onChange' ];
+	};
+
+export type QueryControlsWithMultipleCategorySelectionProps =
+	BaseQueryControlsProps & {
+		/**
+		 * An array of category names, renders a `FormTokenField` component
+		 * when passed in conjunction with `onCategoryChange`.
+		 */
+		categorySuggestions?: { [ categoryName: Entity[ 'name' ] ]: Entity };
+		/**
+		 * The selected categories for the `categorySuggestions`.
+		 */
+		selectedCategories?: Category[];
+		/**
+		 * A function that receives the new category value.
+		 * If this is not specified, the category controls are not included.
+		 */
+		onCategoryChange?: FormTokenFieldProps[ 'onChange' ];
+	};
+
+export type QueryControlsProps =
+	| QueryControlsWithSingleCategorySelectionProps
+	| QueryControlsWithMultipleCategorySelectionProps;
```
</details>

Reply:
Thanks so much for your detailed work, like splitting up `QueryControlsProps` into 2 types.

Committed your diff verbatim in https://github.com/WordPress/gutenberg/compare/6a8920643ec641c27270b9d6c1fb6171131864b1...a90ced1eecce0590c5a644a7879786011ff3f59f